tests/test_core/test_task.py

```diff
@@ -0,0 +1,237 @@
+        t.q.put(resp0)
+        t.q.put(resp2)
+        t.wait_all( f_wait1, 0)
+        self.assertEqual(t._futures, {1: f1, 3: f3})
+        self.assertEqual(f0.done(), True)
+        self.assertEqual(f1.done(), False)
+        self.assertEqual(f2.done(), True)
+        self.assertEqual(f3.done(), False)
+        self.assertEqual(self.proc.q.qsize(), 0)
+
+        resp3 = Response.Delta(3, None, None)
+        t.q.put(resp3)
+        f_wait1 = [f3]
+        self.assertRaises(ValueError, t.wait_all, f_wait1, 0.01)
+        t.stop()
+        self.assertRaises(RuntimeWarning, t.wait_all, f_wait1, 0.01)
+
+        resp1 = Response.Update(1, None, None)
+        resp1.fields["value"] = 'testVal'
+        t.q.put(resp1)
+        self.assertRaises(queue.Empty, t.wait_all, f_wait1, 0.01)
+        self.assertEqual(t._futures, {})
+
+        t._futures = {0: f0, 1: f1, 2: f2}
+        t.q.put(resp1)
+        t.q.put(Task.TASK_STOP)
+        self.assertEqual(f1.result(), 'testVal')
+
+    def test_wait_all_missing_futures(self):
+        # unsolicited response
+        t = Task("testTask", self.proc)
+        f1 = Future(t)
+        resp10 = Response.Return(10, None, None)
+        t.q.put(resp10)
+        t.q.put(Task.TASK_STOP)
+        self.assertRaises(RuntimeWarning, t.wait_all, f1, 0)
+
+        # same future twice
+        f2 = Future(t)
+        t._futures = {1: f2}
+        resp1 = Response.Return(1, None, None)
+        t.q.put(resp1)
+        t.q.put(Task.TASK_STOP)
+        t.wait_all(f2,0)
+        t.wait_all(f2,0)
+
+
+    def _callback(self, value, a, b):
+        self.callback_result = a+b
+        self.callback_value = value
+
+    def test_subscribe(self):
+        t = Task("testTask", self.proc)
+        resp = Response.Update(0, None, None)
+        resp.fields["value"] = 'changedVal'
+        t.q.put(resp)
+        t.stop()
+
+        new_id = t.subscribe(self.attr, self._callback, 3, 5)
+        f1 = Future(t)
+        t._futures = {1: f1}
+
+        self.assertRaises(RuntimeWarning, t.wait_all, f1, 0)
+        self.assertEqual(self.callback_value, 'changedVal')
+        self.assertEqual(self.callback_result, 8)
+        t.unsubscribe(new_id)
+
+    def test_callback_error(self):
+        t = Task("testTask", self.proc)
+        resp = Response.Error(0, None, None)
+        resp.fields["message"] = 'error'
+        t.q.put(resp)
+        t.stop()
+
+        t.subscribe(self.attr, self._callback, 3, 5)
+        f1 = Future(t)
+        t._futures = {1: f1}
+        self.assertRaises(RuntimeError, t.wait_all, f1, 0)
+
+    def test_callback_unexpected(self):
+        t = Task("testTask", self.proc)
+        resp = Response.Delta(0, None, None)
+        t.q.put(resp)
+        t.stop()
+        t.subscribe(self.attr, self._callback, 3, 5)
+        f1 = Future(t)
+        t._futures = {1: f1}
+        self.assertRaises(ValueError, t.wait_all, f1, 0)
+
+
+    def _bad_callback(self, value):
+        raise RuntimeWarning()
```

Comment:
Might want to raise something different here so we can tell what exception is actually returned